rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,241 @@
+
+The DNS Records will have a finalizer from the DNS Operator which will clean the records from the zone before removing the finalizer.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Terminology
+ - zone / DNS Provider zone: The set of records listed in the DNS Provider (the live list of DNS records)
+ - KuadrantRecord: The DNSRecord created on the local cluster by Kuadrant - only contains the DNS requirements for the local cluster
+ - ZoneRecord: The DNSRecord created on the local cluster by the DNS Operator, this reflects the DNSRecord as it appears in the zone.
+
+## General Logic Overview
+
+The general flow for in the Kuadrant operator follows a single path, where it will always output a DNS Record (kuadrantRecord) which specifies everything needed for this workload on this cluster, and is unaware of the concept of distributed DNS, or phrased more simply:
+Build a DNS Record based on the local gateways and DNS Policy.
+
+This kuadrantRecord is reconciled by the DNS Operator, the DNS Operator will act with the DNS Provider's zone as the authority and ensure that the records in the zone relevant to the hosts defined within gateways on it's cluster are accurate. When cleaning up a DNS Record the operator will ensure all records owned by the policy for the gateway on the local cluster are removed and then prune unresolvable records (an unresolvable record is a record that has no logical value such as an IP Address or a CNAME to a different root hostname) related to the hosts defined in the gateway from the DNS Provider zone.
+
+### API Updates
+New configuration options that can be applied as runtime arguments (to allow us emergency configuration rescue in case of an unexpected issue):
+- `RefreshInterval`
+  - When absent it defaults to the shortest TTL of created records (and can be backed off in case of encountering API rate-limits). This is how often in seconds to reconcile the DNS Records in order to pull fresh zone values from the DNS Provider.
+- `MultiClusterFailureThreshold`
+  - When absent defaults to `66`, what percentage of clusters need to be reporting an endpoint unhealthy on a remote IP for it to be removed from the zone by a cluster that does not own that IP.
+- `clusterID`
+  - When absent is generated in code (more information below). A unique string to identify this cluster apart from all other clusters.
+
+New fields added to the DNS Record CRD:
+- `HealthCheck`
+  - Propagated from the DNS Policy, to facilitate reconciling health checks in the DNS controller.
+
+### Changes to DNS Policy reconciler (kuadrant operator)
+
+This will primarily need to be made aware of the new fields on the DNS Policy (Healthcheck block) that it will need to propagate into any DNS Records that it creates or updates. Otherwise the logic of this controller remains largely the same.
```

Comment:
@philbrookes wont the DNSPolicy reconciler need to indicate a change of address also?

Reply:
This could be necessary, though I'd hope we can see the change of IP in the DNS operator without needing to indicate it from the Kuadrant operator.